### ISSUE TITLE
The provided destination instance is not of the required type.

### ISSUE THREAD

**Issue by Kilrath** [modelmapper/modelmapper#443]
Hi, 

I think there is a bug in when try to map a primitive value via global provider. 
The MappingEngineImpl.validateDestination() method does not use Primitives.wrapperFor(destinationType) which produces always following errors:
1) The provided destination instance 9 is not of the required type int.
2) The provided destination instance false is not of the required type boolean.

This error also happens when the primitive source value is directly assigned to destination (destinationType is primitive)!

Can this issue be fixed immediately, please?

**Comment by joeltoby** [comment]
Hi @Kilrath,  Did you ever get a fix or workaround for this issue?  I'm hitting an identical error:

`The provided destination instance true is not of the required type boolean`

**Comment by joeltoby** [comment]
I believe this issue may be linked to issue #450.  Although the exception message (see above) indicates that the value being converted is `true` or `false`, it appears that the `BooleanConverter` is receiving the original Source Object in the context argument instead of the true/false value returned by the provider.

My provider:

Provider<Boolean> isOwnerProvider = req -> {
      boolean isOwner = false;
      
      InstanceOwner ownerEntity = (InstanceOwner) req.getSource();
      
      if(ownerEntity != null && sessionState != null && sessionState.getAuthedUser() != null) {
        if(sessionState.getAuthedUser().getId().equals(ownerEntity.getUserId())) {
          isOwner = true;
        }
      }
      return isOwner;
    };

The provider above expects an instance of `InstanceOwner.class` and returns a boolean.

I have confirmed that my Provider is correctly invoked and that it returns the expected boolean value.  However, after the provider returns its value, `BooleanConverter` is called and my debugger shows that the value of `context.getSource()` is:

"InstanceOwner(id=7c243875-7ada-42f5-8b3e-949aeab4a41e, fullName=Joel Toby .....)"

In other words, the original instance of InstanceOwner.  This matches the behaviour highlighted in #450 I think.

For now I'll have to abandon my Provider in favour of a much less elegant and hacky Converter, but I would be very grateful if you could suggest any workarounds please?

**Comment by joeltoby** [comment]
I have tested with versions 2.3.1 and 2.3.3 - results are the same.

Also tried disabling implicitMapping but this doesn't seem to help.

**Comment by Kilrath** [comment]
Hi,

unfortunately I did the same as you and use a converter if the provider fails due to this issue. 

Sorry

**Comment by joeltoby** [comment]
Thanks for taking the time to reply @Kilrath .

In case others hit the same issue, here is my workaround (using a Converter instead of a Provider):

```
// Disable implicit mappings
ModelMapper modelMapper = new ModelMapper();
modelMapper.getConfiguration().setImplicitMappingEnabled(false);

// Create a converter to provide the value you need
  Converter<InstanceOwner, Boolean> instanceOwnerToIsOwnerConverter = ctx -> {
    boolean isOwner = false;

    if(ctx.getSource() != null) {
      InstanceOwner ownerEntity = (InstanceOwner) ctx.getSource();

      if(ownerEntity != null && sessionState != null && sessionState.getAuthedUser() != null) {
        if(sessionState.getAuthedUser().getId().equals(ownerEntity.getUserId())) {
          isOwner = true;
        }
      }
    }
    return isOwner;
  };


// Create a PropertyMap for the conversion which uses the Converter for one property
private PropertyMap<Item, ItemDTO> ItemToDtoMapping() {
    return new PropertyMap <Item, ItemDTO>() {
      protected void configure() {
        // The false value in `setOwner(false)` is not used
        using(instanceOwnerToIsOwnerConverter).map(source.getHumanOwner()).setOwner(false);
      }
    };
  }

// Finally, add the mapper and add implicitMappings back on
modelMapper.addMappings(ItemToDtoMapping()).implicitMappings();
```
